### PR TITLE
cancel the outgoing when the incoming request got canceled

### PR DESCRIPTION
**Pull request by aryszka**
The current approach relies on using a shared context (context.Context). All consequences of this need yet to be verified, but it's open for testing in deployments.

Signed-off-by: Arpad Ryszka <arpad.ryszka@gmail.com>

### CODE REVIEW

File: dataclients/routestring/example_test.go [hunk 1]
```diff
@@ -39,7 +39,4 @@ func Example() {
 	}
 
 	fmt.Println(string(content))
-
-	// Output:
-	// Hello, world!
```
Comment (szuecs) [inline dataclients/routestring/example_test.go]: Was this on purpose or some linter recommended it?

File: proxy/longpoll_test.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+			}
+		}
+
+		if sendHeader {
+			// in practice, this doesn't block
+			w.WriteHeader(http.StatusOK)
+			w.(interface{ Flush() }).Flush()
+		}
+
+		for loop := true; loop; {
```
Comment (szuecs) [inline proxy/longpoll_test.go]: why do you need `loop` and `for {}` is not enough ?

File: dataclients/routestring/example_test.go (hunk 1 above)
Comment (aryszka) [reply dataclients/routestring/example_test.go]: for some reason the tests failed here too often. It's unrelated from this PR, I should put it back.

File: proxy/longpoll_test.go (hunk 2 above)
Comment (aryszka) [reply proxy/longpoll_test.go]: not needed, at some point this loop had a break in a select, but even then, this is not the right way, it's WIP garbage

**Comment by aryszka** [comment]
eventually, i was able to repro the failing build in the master branch, but seems like it happens more often here. 'Seems', maybe not.

File: dataclients/routestring/example_test.go (hunk 1 above)
Comment (aryszka) [reply dataclients/routestring/example_test.go]: i see now. There is a race condition in the example and we don't know when will skipper return 404 as no routes have been configured yet, or the expected output. We can either make the test less readable by adding the necessary synchronization or we can just skip checking the output.

Comment (szuecs) [reply dataclients/routestring/example_test.go]: fine for me

**Comment by aryszka** [comment]
see this issue and the reference PR: https://github.com/zalando/skipper/pull/857

the underlying case was never proven, or that it is actually a production issue and not only log spam

File: proxy/longpoll_test.go (hunk 2 above)
Comment (aryszka) [reply proxy/longpoll_test.go]: fixed

Comment (szuecs) [reply proxy/longpoll_test.go]: this was not changed

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
thanks for the review!

**Comment by aryszka** [comment]
:+1:

**Pull request merged**